habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -1326,3 +1330,242 @@ def obj_next_to(
+    :param obj_bottom_location: The optional location of the candidate bottom point of the object. If not provided, project the object center to the lowest extent.
+    :param max_dist_to_rec: The threshold point to mesh distance for an object to be matched with a Receptacle.
+    :param candidate_receptacles: Optionally provide a dict (unique_name to Receptacle) of candidate Receptacles for matching. If not provided, uses sim.receptacles.
+    :param island_index: Optionally provide an island_index for identifying navigable floor points. Default is full navmesh.
+    :return: Tuple containing: (1): list of receptacle strings: "floor,region", Receptacle.unique_name, or None (2): a floating point confidence score [0,1] (3): a message string describing the results for use in a UI tooltip
+
+    When using this util for candidate placement with raycasting (e.g. HitL): provide 'support_surface_id' and 'obj_bottom_location' overrides from the raycast.
+    When using this util for assessing current object state (e.g. episode evaluation): leave 'support_surface_id' and 'obj_bottom_location' as default.
+    """
+    # collect information about failure of the function
+    info_text = ""
+
+    # get the center point of the object projected on the local bounding box size in the gravity direction
+    grav_vector = mn.Vector3(0, -1, 0)
+    dist, center = get_obj_size_along(sim, obj.object_id, grav_vector)
+    # either compute or use the provided object location
+    obj_bottom_point = (
+        center + grav_vector * dist
+        if obj_bottom_location is None
+        else obj_bottom_location
+    )
+
+    # find a support surface if one was not provided
+    if support_surface_id is None:
+        # TODO: This hack is the solution to the Bullet physics raycast margin bug. It should be migrated into every raycast from habitat-sim and then removed here.
+        y_buffer = 0.08
+        raycast_results = sim.cast_ray(
+            habitat_sim.geo.Ray(
+                center + mn.Vector3(0, y_buffer, 0), grav_vector
+            )
+        )
+        if raycast_results.has_hits():
+            for hit in raycast_results.hits:
+                if hit.object_id != obj.object_id:
+                    support_surface_id = hit.object_id
+                    # print(f"first_hit true dist = {hit.ray_distance - y_buffer}")
+                    if hit.ray_distance >= y_buffer:
+                        break
+        if support_surface_id is None:
+            info_text = "No support surface found for object."
+            return [], 1.0, info_text
+        # sup_obj = get_obj_from_id(sim,support_surface_id)
+        # print(f"computed support surface {support_surface_id}:{sup_obj.handle if sup_obj is not None else 'stage'}")
+
+    try_floor = False
```

Comment:
I would rename this variable - it's purpose is unclear.

Reply:
Sure, the idea here is that we first try matching to Receptacle objects. If that fails, we fall-back to ensuring the point is navigable (i.e. matching it to the "floor" concept Receptacle.